### PR TITLE
Docker 1.21 API - networks support

### PR DESCRIPTION
**Pull request by marco-gulino**
Implemented the following API for docker 1.9 network management:
 * Create networks
 * Connect containers to networks
 * Disconnect containers to networks
 * List/inspect networks
 * Destroy networks

### CODE REVIEW

File: .checkstyle [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (davidxia) [inline .checkstyle]: Was this file meant to be committed?

**Comment by davidxia** [comment]
@damianopezzotti @marco-gulino  Thanks for the PR, taking a look now.

File: circle.yml [hunk 2]
```diff
@@ -12,6 +12,8 @@ test:
 
 machine:
   pre:
+#    - sudo curl -L -o /usr/bin/docker 'https://s3-external-1.amazonaws.com/circle-downloads/docker-1.9.0-circleci'
```
Comment (davidxia) [inline circle.yml]: Once you rebase, this will go away as it's already in master.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 3]
```diff
@@ -19,14 +19,15 @@
...
-import com.google.common.base.Optional;
-import com.google.common.base.Preconditions;
-import com.google.common.io.CharStreams;
-import com.google.common.net.HostAndPort;
+
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.node.JsonNodeFactory;
+import com.google.common.base.Optional;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Sorry for the nitpick, but could we avoid reordering imports here? It makes the diff misleading.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 4]
```diff
@@ -165,25 +168,29 @@ public void progress(ProgressMessage message) throws DockerException {
   private static final long DEFAULT_READ_TIMEOUT_MILLIS = SECONDS.toMillis(30);
   private static final int DEFAULT_CONNECTION_POOL_SIZE = 100;
 
-  private static final ClientConfig DEFAULT_CONFIG = new ClientConfig(
-      ObjectMapperProvider.class,
-      JacksonFeature.class,
-      LogsResponseReader.class,
-      ProgressResponseReader.class);
+  private static final ClientConfig DEFAULT_CONFIG =
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Perhaps we can revert these code style changes? Looks like it might've been done by your IDE automatically.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 5]
```diff
@@ -165,25 +168,29 @@ public void progress(ProgressMessage message) throws DockerException {
...
-      new GenericType<List<Image>>() {};
+  private static final GenericType<List<Image>> IMAGE_LIST = new GenericType<List<Image>>() {
+  };
+
+  private static final GenericType<List<Network>> NETWORK_LIST = new GenericType<List<Network>>() {
+  };
 
   private static final GenericType<List<ImageSearchResult>> IMAGES_SEARCH_RESULT_LIST =
-      new GenericType<List<ImageSearchResult>>() {};
+      new GenericType<List<ImageSearchResult>>() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Perhaps we can revert these code style changes? Looks like it might've been done by your IDE automatically.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 6]
```diff
@@ -266,14 +274,11 @@ protected DefaultDockerClient(final Builder builder) {
...
-        .build();
-    this.noTimeoutClient = ClientBuilder.newBuilder()
-        .withConfig(config)
+    final RequestConfig noReadTimeoutRequestConfig =
+        RequestConfig.copy(requestConfig).setSocketTimeout((int) NO_TIMEOUT).build();
+    this.noTimeoutClient = ClientBuilder.newBuilder().withConfig(config)
         .property(ApacheClientProperties.CONNECTION_MANAGER, noTimeoutCm)
-        .property(ApacheClientProperties.REQUEST_CONFIG, noReadTimeoutRequestConfig)
-        .build();
+        .property(ApacheClientProperties.REQUEST_CONFIG, noReadTimeoutRequestConfig).build();
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Hm, seems like a lot of these changes are unrelated to the actual network feature. Perhaps we can revert just to make the diff easier to review?

File: src/main/java/com/spotify/docker/client/DockerClient.java [hunk 7]
```diff
@@ -286,9 +286,13 @@ InputStream save(String image, AuthConfig authConfig)
   /**
    * Pull a private docker container image.
    *
-   * @param image The image to pull.
+   * @param image      The image to pull.
    * @param authConfig The authentication config needed to pull the image.
+<<<<<<< HEAD
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DockerClient.java]: woops :(

Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/DockerClient.java]: Hi,
Developer is on holiday until january 4.
We will decide the code next year:-)

Thanks

Il mercoledì 23 dicembre 2015, David Xia <notifications@github.com> ha
scritto:

> In src/main/java/com/spotify/docker/client/DockerClient.java
> <https://github.com/spotify/docker-client/pull/307#discussion_r48367397>:
>
> >     * @param authConfig The authentication config needed to pull the image.
> > +<<<<<<< HEAD
>
> woops :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/307/files#r48367397>.
>


-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

**Comment by damianopezzotti** [comment]
Master imported on branch, unnecessary code changes removed

File: src/main/java/com/spotify/docker/client/messages/Ipam.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class Ipam {
+
+  public Ipam() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Ipam.java]: Looks like we don't need this constructor.

Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/Ipam.java]: removed

2015-12-28 18:20 GMT+01:00 David Xia <notifications@github.com>:

> In src/main/java/com/spotify/docker/client/messages/Ipam.java
> <https://github.com/spotify/docker-client/pull/307#discussion_r48492539>:
>
> > +
> > +import com.fasterxml.jackson.annotation.JsonAutoDetect;
> > +import com.fasterxml.jackson.annotation.JsonProperty;
> > +import com.google.common.base.MoreObjects;
> > +import com.google.common.base.Objects;
> > +
> > +import java.util.ArrayList;
> > +import java.util.List;
> > +
> > +import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
> > +import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
> > +
> > +@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
> > +public class Ipam {
> > +
> > +  public Ipam() {
>
> Looks like we don't need this constructor.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/307/files#r48492539>.
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

File: src/main/java/com/spotify/docker/client/messages/Network.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  @JsonProperty("Driver")
+  private String driver;
+  @JsonProperty("IPAM")
+  private Ipam ipam;
+  @JsonProperty("Containers")
+  Map<String, Container> containers;
+  @JsonProperty("Options")
+  private Map<String, String> options;
+
+  private Network() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Network.java]: Remove?

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class NetworkCreation {
+
+  @JsonProperty("Id") private String id;
+  @JsonProperty("Warnings") private ImmutableList<String> warnings;
+
+  public NetworkCreation() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: unused

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.List;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class NetworkCreation {
+
+  @JsonProperty("Id") private String id;
+  @JsonProperty("Warnings") private ImmutableList<String> warnings;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: Does this work? [Docs here](https://docs.docker.com/engine/reference/api/docker_remote_api_v1.21/#create-a-network) imply it's `Warning` and just a `String`.

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 12]
```diff
@@ -1137,6 +1144,83 @@ public ContainerStats stats(final String containerId)
...
+                Entity.json(request));
+    switch (response.getStatus()) {
+      case 200:
+        return;
+      case 404:
+        throw new ContainerNotFoundException(containerId);
+      case 500:
+        throw new DockerException(response.readEntity(String.class));
+    }
+    System.out.println(response.getStatus());
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Do we need this?

Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: Removed

2015-12-28 18:36 GMT+01:00 David Xia <notifications@github.com>:

> In src/main/java/com/spotify/docker/client/DefaultDockerClient.java
> <https://github.com/spotify/docker-client/pull/307#discussion_r48493435>:
>
> > +    final WebTarget resource = resource().path("networks").path(networkId).path(methodname);
> > +
> > +    Map<String, String> request = new HashMap<>();
> > +    request.put("Container", containerId);
> > +    Response response =
> > +        request(POST, Response.class, resource, resource.request(APPLICATION_JSON_TYPE),
> > +                Entity.json(request));
> > +    switch (response.getStatus()) {
> > +      case 200:
> > +        return;
> > +      case 404:
> > +        throw new ContainerNotFoundException(containerId);
> > +      case 500:
> > +        throw new DockerException(response.readEntity(String.class));
> > +    }
> > +    System.out.println(response.getStatus());
>
> Do we need this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/307/files#r48493435>.
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java (hunk 11 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: Transformed in String

2015-12-28 18:29 GMT+01:00 David Xia <notifications@github.com>:

> In src/main/java/com/spotify/docker/client/messages/NetworkCreation.java
> <https://github.com/spotify/docker-client/pull/307#discussion_r48493055>:
>
> > +
> > +import com.fasterxml.jackson.annotation.JsonAutoDetect;
> > +import com.fasterxml.jackson.annotation.JsonProperty;
> > +import com.google.common.base.MoreObjects;
> > +import com.google.common.collect.ImmutableList;
> > +
> > +import java.util.List;
> > +
> > +import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
> > +import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
> > +
> > +@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
> > +public class NetworkCreation {
> > +
> > +  @JsonProperty("Id") private String id;
> > +  @JsonProperty("Warnings") private ImmutableList<String> warnings;
>
> Does this work? Docs here
> <https://docs.docker.com/engine/reference/api/docker_remote_api_v1.21/#create-a-network>
> imply it's Warning and just a String.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/spotify/docker-client/pull/307/files#r48493055>.
>



-- 
Damiano Pezzotti
...................................... . .
Software Architect | Alambitco
mobile +39 333 2116258
fax +39 02 700432209
email damy@alambitco.com
web www.alambitco.com
flickr www.flickr.com/photos/damianopezzotti
linkedin www.linkedin.com/in/damianopezzotti
pgp id 0x12DFD989

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 13]
```diff
@@ -169,7 +173,7 @@ public void progress(ProgressMessage message) throws DockerException {
       ObjectMapperProvider.class,
       JacksonFeature.class,
       LogsResponseReader.class,
-      ProgressResponseReader.class);
+                       ProgressResponseReader.class);
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: nit: There are still a bunch of random whitespace changes like here. Can we remove these?

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java [hunk 14]
```diff
@@ -218,7 +225,7 @@ public DefaultDockerClient(final URI uri) {
 
   /**
    * Create a new client with default configuration.
-   * @param uri The docker rest api uri.
+   * @param uri                The docker rest api uri.
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: unnecessary whitespace change here

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 15]
```diff
@@ -2000,6 +2004,142 @@ public void testStats() throws DockerException, InterruptedException {
...
+      }
+    }
+    assertThat(network, is(notNullValue()));
+    assertThat(network.id(), is(notNullValue()));
+    assertThat(sut.inspectNetwork(network.id()).name(), is(networkName));
+    assertThat(network.ipam(), equalTo(ipam));
+
+    sut.removeNetwork(network.id());
+
+    try {
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: Would using the `@Rule ExpectedException` above be cleaner here? I.e. `exception.expect(NetworkNotFoundException.class)`.

File: src/main/java/com/spotify/docker/client/messages/NetworkSettings.java [hunk 16]
```diff
@@ -37,9 +37,6 @@
   @JsonProperty("Ports") private Map<String, List<PortBinding>> ports;
   @JsonProperty("MacAddress") private String macAddress;
 
-  private NetworkSettings() {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/NetworkSettings.java]: If you remove this ctor, you should add `@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)` above the class. I think this is why the CircleCI tests are failing with 

```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: No suitable constructor found for type [simple type, class com.spotify.docker.client.messages.NetworkSettings]: can not instantiate from JSON object (missing default constructor or creator, or perhaps need to add/enable type information?)
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@40c55905; line: 1, column: 459] (through reference chain: com.spotify.docker.client.messages.ContainerInfo["NetworkSettings"])
```

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.docker.client.messages;
+
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: unused import

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.util.List;
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: unused import

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 19]
```diff
@@ -2000,6 +2004,142 @@ public void testStats() throws DockerException, InterruptedException {
...
+    assertThat(networkCreation.id(), is(notNullValue()));
+    final ContainerConfig containerConfig =
+        ContainerConfig.builder().image(BUSYBOX_LATEST).cmd("sh", "-c", "echo hello").build();
+    final ContainerCreation containerCreation = sut.createContainer(containerConfig, containerName);
+    assertThat(containerCreation.id(), is(notNullValue()));
+    try {
+      sut.connectToNetwork(containerCreation.id(), networkCreation.id());
+      fail("Expected to throw DockerException");
+    } catch (DockerException e) {
+      assertThat(e.getMessage(), containsString("is not running"));
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: Use `ExpectedException` here too perhaps.

File: src/main/java/com/spotify/docker/client/messages/Config.java [hunk 20]
```diff
@@ -0,0 +1,89 @@
...
+import com.fasterxml.jackson.annotation.JsonAutoDetect;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.ANY;
+import static com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility.NONE;
+
+@JsonAutoDetect(fieldVisibility = ANY, getterVisibility = NONE, setterVisibility = NONE)
+public class Config {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/client/messages/Config.java]: Can we name this class `NetworkConfig` as `Config` seems too generic. We have a `ContainerConfig` class that corresponds to the "Config" JSON key.

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java [hunk 21]
```diff
@@ -2000,6 +2004,142 @@ public void testStats() throws DockerException, InterruptedException {
...
+    assertThat(network.containers().get(containerCreation.id()), notNullValue());
+    sut.stopContainer(containerCreation.id(), 1);
+    sut.removeContainer(containerCreation.id());
+    sut.removeNetwork(networkCreation.id());
+
+  }
+
+
+  @Test
+  public void testNetworksDisconnectContainer() throws Exception {
```
Comment (davidxia) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: Would it make sense to merge this test with the one above? The only additional thing tested is that the container can disconnect from the network.

**Comment by davidxia** [comment]
FYI we use Google's IntelliJ code style guide https://github.com/google/styleguide/. Perhaps that will help with the indentation and whitespaces.

File: src/main/java/com/spotify/docker/client/messages/NetworkSettings.java (hunk 16 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/NetworkSettings.java]: If I remove constructor and add `@JsonAutoDetect`, I have the same exception

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java (hunk 13 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: removed

File: src/main/java/com/spotify/docker/client/DefaultDockerClient.java (hunk 14 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/DefaultDockerClient.java]: removed

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java (hunk 15 above)
Comment (damianopezzotti) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: done

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java (hunk 17 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: removed

File: src/main/java/com/spotify/docker/client/messages/NetworkCreation.java (hunk 18 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/NetworkCreation.java]: removed

File: src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java (hunk 19 above)
Comment (damianopezzotti) [inline src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java]: done

File: src/main/java/com/spotify/docker/client/messages/Config.java (hunk 20 above)
Comment (damianopezzotti) [inline src/main/java/com/spotify/docker/client/messages/Config.java]: NetworkConfig is already present. Renamed in IpamConfig
